src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala

```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
     columnName ~ columnType ^^ {
       case(name ~ ty) => Column(name, ty)
     }
+
+  def CategoricalParser = "Categorical" ~> "(" ~> """\d+""".r <~ ")" ^^ { n => MultinomialType(n.toInt) }
+  def BooleanParser = "Boolean" ^^ { s => BooleanType }
+  def dataType = CategoricalParser | BooleanParser
+
   def schemaDeclaration: Parser[SchemaDeclaration] =
-    relationName ~ opt("?") ~ "(" ~ rep1sep(columnDeclaration, ",") ~ ")" ^^ {
-      case (r ~ isQuery ~ "(" ~ attrs ~ ")") => {
+    relationName ~ opt("?") ~ "(" ~ rep1sep(columnDeclaration, ",") ~ ")" ~ opt(dataType) ^^ {
```

Comment:
How about using a `"::"` between the datatype and relation name with columns?  We're using that for type casting (or type annotation) in the expressions.

Also, only variable relations should be possible to have datatype, right?  Then, how about splitting into two alternative cases?